src/audio/midi/MidiAudio.cpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "MidiAudio.hpp"
+#include "../terminal/parser/stateMachine.hpp"
+
+namespace
+{
+    class MidiOut
+    {
+    public:
+        static constexpr auto NOTE_OFF = 0x80;
+        static constexpr auto NOTE_ON = 0x90;
+        static constexpr auto PROGRAM_CHANGE = 0xC0;
+
+        // We're using a square wave as an approximation of the sound that the
+        // original VT525 terminals might have produced. This is probably not
+        // quite right, but it works reasonably well.
+        static constexpr auto SQUARE_WAVE_SYNTH = 80;
+
+        MidiOut() noexcept
+        {
+            midiOutOpen(&handle, MIDI_MAPPER, NULL, NULL, CALLBACK_NULL);
+            OutputMessage(PROGRAM_CHANGE, SQUARE_WAVE_SYNTH);
+        }
+        ~MidiOut() noexcept
+        {
+            midiOutClose(handle);
+        }
+        void OutputMessage(const int b1, const int b2, const int b3 = 0, const int b4 = 0) noexcept
+        {
+            midiOutShortMsg(handle, MAKELONG(MAKEWORD(b1, b2), MAKEWORD(b3, b4)));
+        }
+
+        MidiOut(const MidiOut&) = delete;
+        MidiOut(MidiOut&&) = delete;
+        MidiOut& operator=(const MidiOut&) = delete;
+        MidiOut& operator=(MidiOut&&) = delete;
+
+    private:
+        HMIDIOUT handle = nullptr;
+    };
+}
+
+using namespace std::chrono_literals;
+
+MidiAudio::~MidiAudio() noexcept
+{
+    try
+    {
+#pragma warning(suppress : 26447)
+        // We acquire the lock here so the class isn't destroyed while in use.
+        // If this throws, we'll catch it, so the C26447 warning is bogus.
+        _inUseMutex.lock();
+    }
+    catch (...)
+    {
+        // If the lock fails, we'll just have to live with the consequences.
+    }
+}
+
+void MidiAudio::Initialize()
+{
+    _shutdownFuture = _shutdownPromise.get_future();
+}
+
+void MidiAudio::Shutdown()
+{
+    // Once the shutdown promise is set, any note that is playing will stop
+    // immediately, and the Unlock call will exit the thread ASAP.
+    _shutdownPromise.set_value();
+}
+
+void MidiAudio::Lock()
+{
+    _inUseMutex.lock();
+}
+
+void MidiAudio::Unlock()
+{
+    // We need to check the shutdown status before releasing the mutex,
+    // because after that the class could be destroyed.
+    const auto shutdownStatus = _shutdownFuture.wait_for(0s);
+    _inUseMutex.unlock();
+    // If the wait didn't timeout, that means the shutdown promise was set,
+    // so we need to exit the thread ASAP by throwing an exception.
+    if (shutdownStatus != std::future_status::timeout)
+    {
+        throw Microsoft::Console::VirtualTerminal::StateMachine::ShutdownException{};

```

Comment:
The problem with using a boolean is we then have to propagate that state up through the entire call stack, including all the following (and then some):

* `ControlCore::_terminalPlayMidiNote`
* `Terminal::PlayMidiNote`
* `AdaptDispatch::PlaySounds`
* `OutputStateMachineEngine::ActionCsiDispatch`
* `StateMachine::_ActionCsiDispatch`
* `StateMachine::ProcessCharacter`
* `StateMachine::ProcessString`
* `ControlCore::_connectionOutputHandler`

But half of those calls already use a boolean return value to signal conpty passthrough, so we'd need to come up with some other mechanism for differentiating between those states. In comparison the exception solution seemed a whole lot easier.

Reply:
Oh, I thought it'd be fine to just do this:
```cpp
void MidiAudio::PlayNote(const int noteNumber, const int velocity, const std::chrono::microseconds duration)
{
    if (_shutdown)
    {
        return;
    }

    // ...
}
```

It would still process the VT, but effectively doing anything with it.

Reply:
A note can last up to about 8 seconds I think. But even if it's only say half a second, you could have a whole buffer of notes still pending that could last for minutes. If we just handled this the way we do other sequences, the UI would hang for that entire duration, because we usually only release the terminal lock once the buffer is empty.

So that's why I've got the terminal temporarily unlocked while the sounds is in progress. But if the app/tab is closed during that period, we can't just carry on processing the rest of the buffer, because the user wouldn't want the remaining 10 minutes of audio to be output after they've closed.

Also at that point, half of the framework would have been destroyed anyway, so any attempt to continue processing output would likely end up in something crashing.

Reply:
> What I'm suggesting would be (as far as I can see) entirely equivalent to your current approach. After Shutdown has been called, all future calls to PlayNote return early and the remaining buffer finishes playing immediately. 

But that's not the way it currently works. When `Shutdown` is called, the exception it triggers in the `Unlock` is like a direct jump to `ControlCore::_connectionOutputHandler`. Once that returns there is zero further processing - there are no future calls to `PlayNote` or anything else. If there were, that would almost certainly kill the app, because everything would already have been deleted.

`Shutdown` is called from the `ControlCore` destructor, which also goes on to destroy the `MidiAudio` class. The only thing keeping it alive at that point is the lock in the destructor, and it'll be able to acquire that as soon as the the `Unlock` method throws. From then on there will be no more `MidiAudio`, no more `ControlCore`, no more `Terminal`, and no more `TermControl`.

That's why I say we need to exit the thread ASAP. I mean like immediately that instant. There's no time for anything else. I know this feels a bit precarious, but I can't see a cleaner way to handle it. What you're proposing would just crash (at least as I understand it).

Reply:
This makes sense for TermControl. Who in `conhost` is going to handle this exception? Or are we going to bubble it up and out as a failure? I know we have a lot of failure states for exceptions coming out of lower tiers in conhost already...

I see that MIDI is shut down during `ConsoleCloseProcessState`; what happens when there are multiple applications connected to one console session and one of them exits (perhaps while a note is playing?) What if it _isn't_ the one playing the note? Are we "deathly serial" here and will only process the exit after we shut down MIDI?

Will another process in the same session stand a chance of _`Initialize`-ing_ MIDI again afterwards?

Reply:
In conhost it's caught in the [`WriteConsoleWImplHelper`](https://github.com/microsoft/terminal/blob/57c3953aca49f68acc480ca27533fdc7aad78b43/src/host/_stream.cpp#L1106) and returns as an `E_ABORT` (the `ShutdownException` derives from `wil::ResultException`).

I'm not sure I understand what you mean by multiple processes in the same session. The MIDI handle itself is a shared resource that stays open for as long as the terminal process is running. The `MidiAudio` class that gets shutdown when a tab is closed is just a thing for managing the note blocking. That can be recreated as often as you like - it's just a field in the`ControlCore` class.

I've tested opening and closing multiple tabs while they're all playing sounds at the same time, so I know that works. I don't know if you mean something else though.

Reply:
Ah, I'm thinking about one conhost session (not in Terminal) with multiple attached processes. I misremembered the purpose of `CloseConsoleProcessState`, and I thought that it would fire _per attached process exit._ You can safely ignore that question now that I know CCPS happens per _console_ exit.

ACK on it being caught in `WriteConsoleWImplHelper`. :smile: